src/pytest_cov/plugin.py

```diff
@@ -328,3 +339,33 @@ def cov(request):
 
 def pytest_configure(config):
     config.addinivalue_line("markers", "no_cover: disable coverage for this test.")
+
+
+class TestContextPlugin(object):
+    def __init__(self, cov):
+        self.cov = cov
+
+    def pytest_runtest_setup(self, item):
+        self.switch_context(item, 'setup')
+
+    def pytest_runtest_teardown(self, item):
+        self.switch_context(item, 'teardown')
+
+    def pytest_runtest_call(self, item):
+        self.switch_context(item, 'run')
+
+    def switch_context(self, item, when):
+        context = "{item.nodeid}|{when}".format(item=item, when=when)
+        self.cov.switch_context(context)
+
+
+def pytest_sessionstart(session):
+    config = session.config
+    cov = config.pluginmanager.get_plugin('_cov')
+    if (
+        config.known_args_namespace.cov_context == 'test'
+        and cov
+        and cov.cov_controller
+        and config.known_args_namespace.cov_source
+    ):
+        config.pluginmanager.register(TestContextPlugin(cov.cov_controller.cov), '_cov_contexts')
```

Comment:
I thought it was easier to add this registration to the existing pytest_sessionstart, since this code is digging into the other plugin's details anyway.

Reply:
Hmm, your comment mentions the old way not working with xdist.  I guess we need to improve the tests then to deal with that case?

Reply:
Hmmmm ... this could be merged into the main plugin's pytest_sessionstart. Previously it was done in the plugin's `__init__` - that's why it didn't work in xdist mode.

Reply:
My test uses `@xdist_params`, and still passed the old way.  I guess the test isn't strict enough yet?  BTW: IRC?